### PR TITLE
Add a button to expand the composer to full screen when the input has 3+ lines

### PR DESCRIPTION
**Pull request by neil-marcellini**
### Details

When a user has typed 3+ lines into the composer (chat text input), show an expand button that will expand the composer to be full screen. The user can also collapse the composer if they like. The composer will automatically collapse when the message is sent. If the composer is full screen and the user deletes lines until there are less than 3, the composer will remain expanded until they press the collapse button.

### Fixed Issues

$ https://github.com/Expensify/App/issues/8423

### Tests / QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] Make sure everything looks ok

#### All platforms:
Sign into NewDot with any account

- [ ] **A**

1. Go to a chat and enter 3 lines into the composer (text input)
2. Make sure that the expand button shows up when you enter the 3rd line

- [ ] **B**

1. Delete a line
2. Verify that the expand button is hidden

- [ ] **C**

1. Add a line
2. Click the expand button
3. Make sure that the composer goes full screen
4. Click the collapse button
5. Make sure that the composer collapses back to it's original size.

- [ ] **D**

1. Expand the composer
2. Delete until there is 1 line
3. Verify that the composer is still expanded
4. collapse the composer
5. Verify that the expand button is not shown

- [ ] **E**

1. Paste a lot of text into the composer (this PR description works)
2. Verify that the composer is not full screen
3. Expand the composer
4. Verify that the text fills the full composer
5. Collapse the composer
6. Verify that the composer is not full screen

- [ ] **F**

1. Expand the composer
2. Send the message
3. Verify that the composer collapses after sending

- [ ] **G**

1. Enter 3+ lines in the composer
2. Make sure you can hit the emoji button and select and emoji
3. Expand the composer
4. Make sure you can hit the emoji button and select and emoji

#### Native Mobile only:

- [ ] **Mobile A**

1. Enter 3+ lines into the composer and ensure the keyboard shows
2. Expand the composer
3. Dismiss the keyboard by swiping it down
4. Verify that the composer expands downward to fill the whole screen
5. Collapse the composer
6. Verify that it collapses at the bottom of the screen.

#### Contributor (PR Author) Checklist
- [X] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

#### Web
https://user-images.githubusercontent.com/26260477/171524792-e5d458c4-59ae-4004-8fad-c6b8bd7f4304.mp4

#### Mobile Web

**Android / Chrome**

https://user-images.githubusercontent.com/26260477/171698079-18f021fa-f88e-4199-b433-3f8b7b371d74.mp4

**iOS / Safari **

https://user-images.githubusercontent.com/26260477/171969631-c3bebbb6-faff-4edb-971c-29b3ddc486c8.mp4

#### Desktop

https://user-images.githubusercontent.com/26260477/171969602-fa7e0fdb-66b5-49de-880b-e81e1e0aaffc.mp4

#### iOS

https://user-images.githubusercontent.com/26260477/171524025-2a39cb26-89d1-4403-b3ee-a24cc48908d9.mp4

#### Android

https://user-images.githubusercontent.com/26260477/171959669-329c9ae4-16a3-4a15-afd4-96246b6e7b65.mp4

### CODE REVIEW

**Comment by neil-marcellini** [comment]
Currently, expanding and collapsing the composer only works after refreshing the page, so I'll focus on fixing that. Please let me know what you think of my solution so far.

**Comment by neil-marcellini** [comment]
Expanding and collapsing the composer works now. Next I want to automatically collapse the composer when a message is sent. Then I need to fix a bug where the composer doesn't reduce size when deleting lines from 3 to 0.

**Comment by neil-marcellini** [comment]
iOS and Web are working well right now. Next I'll add support for Android and then I'll check to make sure that the other uses of the Composer (editing, emojis) still function correctly.

**Review by Luke9389 — CHANGES_REQUESTED** [review]

File: src/components/Composer/index.js
Comment (neil-marcellini) [reply src/components/Composer/index.js:96]: I think we need a default because the `Composer` is also used by the emoji picker and the message editor.

**Comment by neil-marcellini** [comment]
I think this is pretty much ready to go, I just need to get some spanish translations done which will fix Travis. Also, currently if you open the full composer and close it you will lose your scroll position in the chat. I'm pretty sure that happens because I'm conditionally rendering the ReportActionsList.

Is it ok if I leave the composer story unchanged?

**Comment by neil-marcellini** [comment]
On Friday I discovered that there are some problems on mobile web (iOS Safari). When you tap into the composer, enter 3+ lines, and expand it, the composer expands past the top of the screen. When the keyboard is displayed, "Safari doesn’t resize the browser window but instead moves it upward such that it is partially offscreen". [Here](https://blog.opendigerati.com/the-eccentric-ways-of-ios-safari-with-the-keyboard-b5aa3f34228d) is a good article about that.

I'm working on a solution where I will use the [visual viewport API](https://developer.mozilla.org/en-US/docs/Web/API/Visual_Viewport_API) to get the viewport height and set the composer to the fill the full visible height. Implementing this solution has been tricky because I only want it to run on mobile web, and it needs to override all of the flexbox styles that normally make the composer full size. [Here](https://allthingssmitty.com/2020/05/11/css-fix-for-100vh-in-mobile-webkit/) is an alternate solution I will explore later that uses `-webkit-fill-available`.

**Comment by neil-marcellini** [comment]
I finally got the full composer to work on iOS Safari. I'll keep uploading screen recordings and I'll fix the merge conflict tomorrow, and then it will be ready for review.

**Comment by neil-marcellini** [comment]
The merge conflicts have to do with changes that are a regression. There is a PR to fix it here https://github.com/Expensify/App/pull/9276, so I will wait until that is merged to fix the conflicts.

**Comment by neil-marcellini** [comment]
I spent a lot of time trying to figure out the vertical spacing for the iOS full composer. I don't like how it cuts off earlier than it does on web. The problem is that the text input doesn't change height in the same way when it has `align-self: flex-end`, but this is needed to have the input start on the bottom.

However, I think it's probably ok as is.
image

File: src/components/Composer/index.native.js
Comment (Luke9389) [inline src/components/Composer/index.native.js:142]: I think we try to stay away from `Platform.OS`.

**Review by Luke9389 — DISMISSED** [review]
This is a first pass. I'll come back for more review comments later this eve or tomorrow.

File: src/components/Composer/index.js
Comment (neil-marcellini) [reply src/components/Composer/index.js:367]: Yeah, great point! Idk what I was thinking there.

File: src/components/Composer/index.native.js
Comment (neil-marcellini) [reply src/components/Composer/index.native.js:142]: I think `Platform.OS` is worth it here, because otherwise I would have a complete duplicate file where the only difference is not passing this prop. I decided to combine the iOS and Android implementations since I would be adding more functionality to this class.

If you feel strongly that this is a bad approach then I could refactor so that we once again have separate implementations, and I can try to extract the shared functions into a utility file.

File: src/components/Composer/index.native.js
Comment (neil-marcellini) [reply src/components/Composer/index.native.js:138]: Good point. That's not my comment but I'll fix it anyways.

File: src/components/Composer/index.js
Comment (Luke9389) [reply src/components/Composer/index.js]: No worries! Whenever I write new files I start with the absolute _most_ readable code possible and then end up making this type of change when I feel done.

File: src/components/Composer/index.native.js
Comment (Luke9389) [reply src/components/Composer/index.native.js]: Awesome!

File: src/components/Composer/index.native.js
Comment (Luke9389) [reply src/components/Composer/index.native.js:105]: OK I did find [this](https://stackoverflow.com/questions/38386704/react-native-determine-number-of-lines-of-text-component).... Maybe theres a `.lines` array we can access? 

Don't spend too much time on this, just a thought.

Comment (neil-marcellini) [reply src/components/Composer/index.native.js:105]: That is for the `<Text>` component, not a `<TextInput>`. I think the approach we have is the best way to do it for now.

File: src/components/Composer/index.native.js
Comment (Luke9389) [reply src/components/Composer/index.native.js:129]: Yea, it is unfortunate that there would be so much overlap... 
Perhaps there's a different way to accomplish this.

It's not really a matter of feeling strongly or not. We just don't use Platform.OS as a matter of policy.

Comment (neil-marcellini) [reply src/components/Composer/index.native.js:129]: Ok. I am creating separate files for ios and android again. I'm moving `updateNumberOfLines` and `getNumberOfLines` into `ComposerUtils/index.native.js` for the mobile implementations and I'll move the web / desktop implementations of those methods into `ComposerUtils/index.js`.

**Comment by neil-marcellini** [comment]
Updated to remove `Platform.OS`.

File: src/components/Composer/index.native.js
Comment (Luke9389) [reply src/components/Composer/index.native.js]: Yup! I think that's a great model for how to do this. Thanks @rushatgabhane

Comment (neil-marcellini) [reply src/components/Composer/index.native.js]: The `BaseModal` holds shared functionality for all platforms. In this case, native is quite different from web so I think the way I refactored makes sense. Please let me know what you think of my latest changes.

**Comment by neil-marcellini** [comment]
Please let me know what you think of my latest changes. @stitesExpensify would you please review when you have a moment?

**Comment by neil-marcellini** [comment]
Updated again to fix merge conflicts. @stitesExpensify could you please review before the end of the day?

**Comment by stitesExpensify** [comment]
Reviewing now

File: src/components/Composer/index.js
Comment (stitesExpensify) [inline src/components/Composer/index.js:332]: Just to make sure I understand, we're setting the number of lines to 1, and then once that's done we're updating the number of lines again to be correct?

File: src/pages/home/report/ReportActionCompose.js
Comment (stitesExpensify) [inline src/pages/home/report/ReportActionCompose.js:149]: Should we make constants for these?

File: src/pages/home/report/ReportActionCompose.js
Comment (stitesExpensify) [inline src/pages/home/report/ReportActionCompose.js:302]: another place we could use constants

File: src/pages/home/report/ReportActionCompose.js
Comment (stitesExpensify) [inline src/pages/home/report/ReportActionCompose.js:515]: This should be a ternary instead of `this.props.isComposerFullSize` and `!this.props.isComposerFullSize`. e.g.

this.props.isComposerFullSize ? (<components>) : (<otherComponents)

**Review by stitesExpensify — COMMENTED** [review]
Looks good! Just a couple of minor nitpicks!

File: src/components/Composer/index.js
Comment (neil-marcellini) [reply src/components/Composer/index.js:332]: Yeah. For some reason we have to set the number of lines to 1 to properly compute the scrollHeight, and I found that I also had to remove the expand / collapse button.

File: src/pages/home/report/ReportActionCompose.js
Comment (neil-marcellini) [reply src/pages/home/report/ReportActionCompose.js:149]: Yeah that sounds like a good idea, I'll add that.

File: src/pages/home/report/ReportActionCompose.js
Comment (neil-marcellini) [reply src/pages/home/report/ReportActionCompose.js:515]: Hmm, but what about this condition `(!this.props.isComposerFullSize && this.state.isFullComposerAvailable)`. With a ternary I would still need to check `this.state.isFullComposerAvailable`, so I kind of think this version is more readable.

**Comment by neil-marcellini** [comment]
Thanks so much for your review! I'm ready for another.

**Review by stitesExpensify — APPROVED** [review]
LGTM! Thanks for the changes :)

**Review by PauloGasparSv — APPROVED** [review]
LGTM!

File: src/pages/home/ReportScreen.js
Comment (marcaaron) [inline src/pages/home/ReportScreen.js:131]: @neil-marcellini nice work on this PR and feature. I came across this use of `window` in a cross-platform component and wanted to call out [this section of the README](https://github.com/expensify/app#platform-specific-file-extensions:~:text=Note%20that%20index.js%20should%20be%20the%20default%20and%20only%20platform%2Dspecific%20implementations%20should%20be%20done%20in%20their%20respective%20files).

We really shouldn't be seeing `window` unless we are in an `index.js` file and there's an `index.native.js` that is doing a no-op.

Comment (neil-marcellini) [reply src/pages/home/ReportScreen.js:131]: Ok, got it. Is it worth making an `index.native.js` version of this component to fix this? Or could I make a `withViewportOffsetTop` HOC that would have a web and native version?

Comment (marcaaron) [reply src/pages/home/ReportScreen.js:131]: Both could work. Though, this component is pretty important and having two might get confusing. I would create a small lib that returns something like `VisualViewport` we can call `addEventListener()` on and for mobile it would just return an empty function.

Comment (marcaaron) [reply src/pages/home/ReportScreen.js:131]: Could also return a cleanup method something like 

**index.js**
```js
function addResizeListener(callback) {
    if (!window.visualViewport) {
        return () => {};
    }

    window.visualViewport.addEventListener('resize', callback);
    return () => window.visualViewport.removeEventListener('resize', callback);
}
```

**index.native.js**
```js
function addResizeListener(callback) {
    return () => {};
}
```

I feel like I've used this pattern before elsewhere to get around the cross-platform guidance.

**Comment by trjExpensify** [comment]
👋 Heads up! As per the BugZero checklist, commenting to surface that this introduced a regression on Android outlined [here](https://github.com/Expensify/App/issues/10202#issuecomment-1307930393). 

CC: @Julesssss @Santhosh-Sellavel @fedirjh

**Comment by neil-marcellini** [comment]
Thanks for the heads up. From the videos in the PR I'm pretty sure everything was aligned when this was merged, but I guess my changes were fragile.

**Pull request merged**
